### DISCUSSION TITLE
Removinf freeze_param from hubert

### DISCUSSION THREAD

**Discussion by pratikkumar018** [espnet/espnet#3790] [Q&A]
We are trying to run https://github.com/espnet/espnet/blob/master/egs2/wsj/asr1/conf/tuning/train_asr_conformer_s3prlfrontend_hubert.yaml config file, but I dont want the frontend to be frozen. So, I have deleted freeze_param: ["frontend.upstream"] line. 
Now it is giving me error in training .    

>  if self.reducer._rebuild_buckets():
> RuntimeError: Expected to have finished reduction in the prior iteration before starting a new one. This error indicates that your module has parameters that were not used in producing loss. You can enable unused parameter detection by (1) passing the keyword argument `find_unused_parameters=True` to `torch.nn.parallel.DistributedDataParallel`; (2) making sure all `forward` function outputs participate in calculating loss. If you already have done the above two steps, then the distributed data parallel module wasn't able to locate the output tensors in the return value of your module's `forward` function. Please include the loss function and the structure of the return value of `forward` of your module when reporting this issue (e.g. list, dict, iterable).

Can you please help.

**Comment by simpleoier** [comment]
Hi @pratikkumar018 , thanks for reporting this. The reason is because in the current implementation, we only considered not to train the frontend. If you want to train it, you also have to fix these two lines here: 
https://github.com/espnet/espnet/blob/25d441dc375f96f0aa8b6189f38e3028ae2aaaa0/espnet2/asr/frontend/s3prl.py#L122
https://github.com/espnet/espnet/blob/25d441dc375f96f0aa8b6189f38e3028ae2aaaa0/espnet2/asr/frontend/s3prl.py#L123

**Reply by kamo-naoyuki** [reply]
@simpleoier `torch.no_grad` has no effect if parameters are freezed. If there are no parameters in the computation, pytorch doesn't make gradients automatically, and in this case the input data is a tensor from data-loader and modules are freezed.

https://github.com/espnet/espnet/blob/25d441dc375f96f0aa8b6189f38e3028ae2aaaa0/espnet2/asr/frontend/s3prl.py#L123

How about removing it and changing `eval()` to

```
if not has_parameters(self.upstream):
    self.upstream.eval() 
```

**Reply by simpleoier** [reply]
@kamo-naoyuki Thanks for your suggestion! I'm going to make a PR to enable training s3prl frontend according to your comment.
